src/mito2/src/worker/handle_drop.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Handling drop request.
+
+use std::time::Duration;
+
+use common_query::Output;
+use common_telemetry::info;
+use common_telemetry::tracing::warn;
+use futures::StreamExt;
+use object_store::util::join_path;
+use object_store::ObjectStore;
+use snafu::ResultExt;
+use store_api::storage::RegionId;
+use tokio::time::sleep;
+
+use crate::error::{OpenDalSnafu, RegionNotFoundSnafu, Result};
+use crate::region::RegionMapRef;
+use crate::worker::{RegionWorkerLoop, DROPPING_MARKER_FILE};
+
+const GC_TASK_INTERVAL_SEC: u64 = 5 * 60;
+
+impl<S> RegionWorkerLoop<S> {
+    pub(crate) async fn handle_drop_request(&mut self, region_id: RegionId) -> Result<Output> {
+        let Some(region) = self.regions.get_region(region_id) else {
+            return RegionNotFoundSnafu { region_id }.fail();
+        };
+
+        info!("Try to drop region {}", region_id);
+        region.stop().await?;
+
+        // write dropping marker
+        let marker_path = join_path(&region.region_dir, DROPPING_MARKER_FILE);
+        self.object_store
+            .write(&marker_path, vec![])
+            .await
+            .context(OpenDalSnafu)?;
+
+        // remove this region from region map to prevent other requests from accessing this region
+        self.regions.remove_region(region_id);
```

Comment:
Why not remove directly in [the first line](https://github.com/GreptimeTeam/greptimedb/pull/2286/files#diff-fcc370977307572f8c0bdeb82ccaf60d9058aea8484f3d39cd8e439b0d2511a4R37) of the method?

Reply:
This line modifies the memory state. It should happen after the disk change (marker file) is made. Otherwise if the disk change fails this region is lost.